doc/rtd/topics/debugging.rst

```diff
@@ -163,3 +163,92 @@ commandline:
 
 Inspect cloud-init.log for output of what operations were performed as a
 result.
+
+.. _proposed_sru_testing:
+
+Stable Release Updates (SRU) testing for cloud-init
+==============================================================
+Once an Ubuntu release has been published, updates to it must follow a special
+procedure called a "stable release update" or `SRU`_.
+
+The cloud-init project has an exception process it follows when validating
+cloud-init for an SRU listed at `CloudinitUpdates`_.
+
+Generally SRU tests of cloud-init perform the following:
+
+ * Install a pre-release version of cloud-init from the
+   **(xenial|bionic|etc)-proposed** APT pocket
+ * Upgrade cloud-init and attempt a clean run of cloud-init to assert the new
+   version of cloud-init works properly on various platforms or Ubuntu series
+ * Check for tracebacks, errors in behavior or performance regressions
+
+
+Manual SRU verification procedure
+---------------------------------
+Below are steps that can manually test cloud-init **-proposed** version
+
+1. Launch a VM on your favorite platform, providing the cloud-config
+   user-data and replacing `<YOUR_LAUNCHPAD_USERNAME>` with your username:
+
+.. code-block:: yaml
+
+    ## template: jinja
+    #cloud-config
+    ssh_import_id: [<YOUR_LAUNCHPAD_USERNAME>]
+    hostname: SRU-worked-{{v1.cloud_name}}
+
+2. Wait for current cloud-init to complete, replace `<YOUR_VM_IP>` with the IP
+   address of the VM that was launched in step 1:
+
+.. code:: bash
+
+    CI_VM_IP=<YOUR_VM_IP>
+    ssh ubuntu@$CI_VM_IP -- cloud-init status --wait
+
+3. Setup **-proposed** pockets on your VM and install **-proposed** cloud-init
+
+.. code:: bash
+
+    # Create a script that will add a -proposed pocket to apt-sources
+    # and install cloud-init from that pocket
+    cat > setup_proposed.sh <<EOF
+    #/bin/bash
+    mirror=http://archive.ubuntu.com/ubuntu
+    echo deb \$mirror \$(lsb_release -sc)-proposed main | tee \
+        /etc/apt/sources.list.d/proposed.list
+    apt-get update -q
+    apt-get install -qy cloud-init
+    EOF
+
+    scp setup_proposed.sh ubuntu@$CI_VM_IP:.
+    ssh ubuntu@$CI_VM_IP -- sudo bash setup_proposed.sh
+
+4. Change hostname, clean cloud-init's state, and reboot to run cloud-init
+   from scratch:
+
+.. code:: bash
+
+    ssh ubuntu@$CI_VM_IP -- sudo hostname something-else
+    ssh ubuntu@$CI_VM_IP -- sudo cloud-init clean --logs --reboot
+
+5. Validate **-proposed** cloud-init came up without error
+
+.. code:: bash
+
+    # Block until cloud-init completes and verify from --long that the proper
```

Comment:
How will the tester to know what the "proper" DS is? For the big clouds, it will be obvious, but we're also already doing testing there. The people following this doc are more likely to be testing a smaller cloud which might not have its own datasource (and could even be using one named after a different cloud, if they are EC2-compatible).

Perhaps we should ask them to make a note of what's used pre-reboot for comparison?

(Am I over-thinking this?)

Reply:
+1 added a comment to step 1
```
    # Make note of the datasource cloud-init detected in --long output.
    # In step 5, confirm the same datasource got detected after upgrade.
    ssh ubuntu@$CI_VM_IP -- cloud-init status --wait --long
```

Reply:
I can't propose across the line break, but can we do something like `s/proper datasource/datasource from step 1/`?